### ISSUE TITLE
[New Debugger] Impossible to step into <debuggerCompleteToSender> methods

### ISSUE THREAD

**Issue by dionisiydk** [guillep/pharo-core#8544]
**Describe the bug**
With new debugger it is impossible to debug the implementation of low level methods marked with <debuggerCompleteToSender>.
For example debug "self error: 'test'" expression and step into #error: method. The debugger will not show the callee context and it will stay like nothing happens.

**Expected behavior**
StepInto should always allow to dive into low level parts.

Originally #debuggerCompleteToSender pragma was introduced to skip the too low level parts of stack when debugger just shows (like to not show Halt machinery when "self halt" is triggered).
But now it seems that the new debugger applies such filters all the time for every step/update operation. That's wrong.

**Comment by dionisiydk** [comment]
Also StepOver has same issue. But maybe here it is intentional.
Anyway such "aggressive" filtering leads to quite different experience comparing to old debuggers. Normally StepOver would show the signaler context of the error. 
Now for example when you Step Over the DNU message the debugger just stops in place. It is not easy to notice the  indication on the window title. And the first impression is that StepOver did nothing.  
Maybe I should just get in used into it. But StepInto is definitely working wrong.

**Comment by StevenCostiou** [comment]
I added an option to unlock that, you can find it in settings https://github.com/pharo-spec/NewTools/issues/140
But it is in NewTools and it is not merged yet.

I will try to see if it behaves as expected using that setting.

**Comment by StevenCostiou** [comment]
![Screenshot 2021-02-10 at 14 01 22](https://user-images.githubusercontent.com/26929529/107513416-77971580-6ba8-11eb-8181-62f4935ccda3.png)

Ok the step in works as you describe with the setting "filter stack" disabled.

I agree it may be annoying to have to go to the settings (there is now a shortcut to get directly on debugger settings from the debugger), but stepping into these low level methods should not interest users that much and interest core developers more.
For now this is the solution I propose, such user experience problem is hard to solve. I am open to improvement suggestions and contributions ;)

@estebanlm can we envision a NewTools merge in the near future?

**Comment by dionisiydk** [comment]
>> I agree it may be annoying to have to go to the settings (there is now a shortcut to get directly on debugger  settings from the debugger), but stepping into these low level methods should not interest users that much and interest core developers more.

I disagree with that. 
StepInto is very explicit command: user wants to go deeply. If he does not want to go deeply he would press StepOver.

If there is a settings I can of course live with that. But think the new behavior is not complete. It does not feel like "low level step" like a step into assignment for example. I must click StepInto two times to actually bypass to the next operation:
```Smalltalk
self assert: true.
1+2
```
Two clicks are needed to go into second line. Compare it with:
```Smalltalk
a := 100.
1+2
```
Another moment. New StepInto behavior does not reflect the process stack. Try again with "assert" and check what is the #suspendedContext of the process. It will be the #assert: method. But Debugger does not show that. I believe that debugger must show the actual process state.
So to make real "low level steps" the StepInto must complete the step, not just a classic StepInto with filtering.

Also think about confusion. Now step into "self error: 'test'" does nothing. No error, no visual change. But the second click will fail (with bad behavior but it is a different issue).

**Comment by dionisiydk** [comment]
It would be nice to keep this issue open. To allow others notice and comment on it.
Because for me the current state is degradation.

**Comment by dionisiydk** [comment]
> So to make real "low level steps" the StepInto must complete the step, not just a classic StepInto with filtering.

Let me rephrase it. Debugger should use StepOver to implement StepInto for low level methods. It is not correct to just hide the stack produced by actual StepInto command.

**Comment by StevenCostiou** [comment]
Denis, basically I do not disagree with your points. I reopened the issue, now there are several possibilities:
- completely disable the filtering and just take care to open the debugger with the first non  `<debuggerCompleteToSender>` context
- keep the filtering when opening the debugger but not when stepping
- remove all filtering by default and keep the setting to activate it

Do you think of other possibilites?

I can do any of these three options in a few minutes, but I agree that the community should comment.

**Comment by dionisiydk** [comment]
Hi Steven.

> Denis, basically I do not disagree with your points. I reopened the issue, now there are several possibilities:
> 
> * completely disable the filtering and just take care to open the debugger with the first non  `<debuggerCompleteToSender>` context

Is it how GT debugger was working?

> * keep the filtering when opening the debugger but not when stepping

And what is the difference here? 
Does it mean that the debugger will not touch the stack at all and will just filter (will not show) the "low level contexts"?
I would prefer this way because opening the debugger will not modify the process. I like to see "self halt" when debugger opens (without Halt machinery) but it is a bit scary to realize that the debugger unwinds the stack behind the scene. And I remember it caused some problems in past. 

But the question is how to present the filtering in the UI. Because simple filter leads to the confusion between real state of the process and what we see in the debugger. I think it is very important to have an indication that there are more details to see and to have a way to expand the view into those invisible parts (like expand button in the tree node but I do not know if it is applicable here)

**Comment by StevenCostiou** [comment]
It is currently only a UI filter for most of the cases: the contexts are just hidden.

To my knowledge only halts do change the stack, I did not check but I believe it is only when you step over a halt:
```Smalltalk
completeProcess: aProcess with: aContext
	^ aContext methodNode
		pragmaNamed: #debuggerCompleteToSender
		ifPresent: [ aProcess completeTo: aContext sender ]
		ifAbsent: [ super completeProcess: aProcess with: aContext ]
```
So a proposition could be:

- When opening the debugger, filter the stack
- Step-in: no filter
- Step-over: filter
- Step-through: ?
- Somewhere an explicit mention that the stack is filtered and a button to unfilter (now I don't do UI design so I'd need help on that)
- In the settings, ways to configure (always filter, never filter...) with always by default

What do you think?

**Comment by dionisiydk** [comment]
I found more problems.
Try to put halt into any of <debuggerCompleteToSender> methods and call it: 
- the debugger will not show the method with halt (due to the filtering I guess)

Maybe simplest would be to disable filtering at all because it breaks too many scenarios in the current form.

**Comment by StevenCostiou** [comment]
I don't get it there is something missing in your text?
Do you mean "a halt in a `<debuggerCompleteToSender>` method?

About disabling filtering: I would prefer to identify and fix the broken scenarios, as the setting will allow anyone to disable it (once NewTools is merged).

**Comment by dionisiydk** [comment]
For example put halt in #should: method:
```Smalltalk
TestAsserter>>should: aBlock
	"Evaluate te Block first then execute the assertion"
	<debuggerCompleteToSender>
	self halt.
	self assert: aBlock value
```
Now trigger it:
```Smalltalk
TestCase new should: true
```
The debugger will not show the #should method.
I believe that this is critical issue that halt does not work correctly everywhere.

**Comment by StevenCostiou** [comment]
ah, I see. 
The halt works, this is just a UI filtering issue.
